File: app/src/main/java/io/github/summerdez/asmrplayer/domain/model/TrackItem.java
package io.github.summerdez.asmrplayer.domain.model;

import io.github.summerdez.asmrplayer.R;
import io.github.summerdez.asmrplayer.data.*;
import io.github.summerdez.asmrplayer.data.remote.*;
import io.github.summerdez.asmrplayer.data.download.*;
import io.github.summerdez.asmrplayer.data.files.*;
import io.github.summerdez.asmrplayer.domain.*;
import io.github.summerdez.asmrplayer.domain.model.*;
import io.github.summerdez.asmrplayer.playback.*;
import io.github.summerdez.asmrplayer.presentation.*;
import io.github.summerdez.asmrplayer.ui.*;
import io.github.summerdez.asmrplayer.ui.activity.*;
import io.github.summerdez.asmrplayer.ui.components.*;
import io.github.summerdez.asmrplayer.ui.screens.*;
import io.github.summerdez.asmrplayer.ui.theme.*;
import io.github.summerdez.asmrplayer.ui.util.*;
import io.github.summerdez.asmrplayer.di.*;
import android.net.Uri;

import org.json.JSONException;
import org.json.JSONObject;

public final class TrackItem {
    public final String id;
    public String title;
    public final String uri;
    public String subtitleUri;
    public String subtitleTitle;
    public final long durationMs;

    public TrackItem(String id, String title, String uri) {
        this(id, title, uri, "", "");
    }

    public TrackItem(String id, String title, String uri, String subtitleUri, String subtitleTitle) {
        this(id, title, uri, subtitleUri, subtitleTitle, 0L);
    }

    public TrackItem(String id, String title, String uri, String subtitleUri, String subtitleTitle, long durationMs) {
        this.id = id;
        this.title = title == null ? "" : title;
        this.uri = uri == null ? "" : uri;
        this.subtitleUri = subtitleUri == null ? "" : subtitleUri;
        this.subtitleTitle = subtitleTitle == null ? "" : subtitleTitle;
        this.durationMs = Math.max(0L, durationMs);
    }

    public boolean hasAudioUri() {
        return !uri.isEmpty();
    }

    public Uri audioUri() {
        return Uri.parse(uri);
    }

    public Uri subtitleUriOrNull() {
        return subtitleUri.isEmpty() ? null : Uri.parse(subtitleUri);
    }

    public String subtitleTitleOr(String fallback) {
        return subtitleTitle.isEmpty() ? fallback : subtitleTitle;
    }

    public JSONObject toJson() throws JSONException {
        JSONObject object = new JSONObject();
        object.put("id", id);
        object.put("title", title);
        object.put("uri", uri);
        object.put("subtitleUri", subtitleUri);
        object.put("subtitleTitle", subtitleTitle);
        object.put("durationMs", durationMs);
        return object;
    }

    public static TrackItem fromJson(JSONObject object) throws JSONException {
        return new TrackItem(
                object.optString("id"),
                object.optString("title"),
                object.optString("uri"),
                object.optString("subtitleUri"),
                object.optString("subtitleTitle"),
                object.optLong("durationMs", 0L));
    }
}

File: app/src/main/java/io/github/summerdez/asmrplayer/domain/model/TrackItem.kt
package io.github.summerdez.asmrplayer.domain.model

import android.net.Uri
import org.json.JSONException
import org.json.JSONObject

class TrackItem(
    val id: String = "",
    var title: String = "",
    val uri: String = "",
    var subtitleUri: String = "",
    var subtitleTitle: String = "",
    durationMs: Long = 0L,
) {
    val durationMs: Long = maxOf(0L, durationMs)

    fun hasAudioUri(): Boolean = uri.isNotEmpty()

    fun audioUri(): Uri = Uri.parse(uri)

    fun subtitleUriOrNull(): Uri? = if (subtitleUri.isEmpty()) null else Uri.parse(subtitleUri)

    fun subtitleTitleOr(fallback: String): String = if (subtitleTitle.isEmpty()) fallback else subtitleTitle

    @Throws(JSONException::class)
    fun toJson(): JSONObject {
        val obj = JSONObject()
        obj.put("id", id)
        obj.put("title", title)
        obj.put("uri", uri)
        obj.put("subtitleUri", subtitleUri)
        obj.put("subtitleTitle", subtitleTitle)
        obj.put("durationMs", durationMs)
        return obj
    }

    companion object {
        @Throws(JSONException::class)
        fun fromJson(obj: JSONObject): TrackItem {
            return TrackItem(
                obj.optString("id"),
                obj.optString("title"),
                obj.optString("uri"),
                obj.optString("subtitleUri"),
                obj.optString("subtitleTitle"),
                obj.optLong("durationMs", 0L),
            )
        }
    }
}
